### PR TITLE
feat: add scrapy_crawler

### LINKED ISSUE #58: add a web crawler to toolkit

### PR DESCRIPTION
**Pull request by ayoubmrx**
closes #58

### CODE REVIEW

File: hyko_toolkit/functions/web/web_scraping/scrapy_crawler/metadata.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    )
+    stop_urls: str = Field(
+        ...,
+        description="A list of urls to stop crawling. Protocol must be either 'http' or 'https'.",
+    )
+
+
+@func.set_param
+class Params(CoreModel):
+    allowed_domains: str = Field(..., description="A domain to crawl.")
```
Comment (1hachem) [inline hyko_toolkit/functions/web/web_scraping/scrapy_crawler/metadata.py:29]: is this one domain or a list of domains ? can this be optional and instead have depth of search ?

File: hyko_toolkit/functions/web/web_scraping/scrapy_crawler/metadata.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    task="web_scraping",
+    description="use scrapy to crawl web pages",
+    absolute_dockerfile_path="./toolkit/hyko_toolkit/functions/web/web_scraping/scrapy_crawler/Dockerfile",
+    docker_context="./toolkit/hyko_toolkit/functions/web/web_scraping/scrapy_crawler",
+)
+
+
+@func.set_input
+class Inputs(CoreModel):
+    start_urls: str = Field(
```
Comment (1hachem) [inline hyko_toolkit/functions/web/web_scraping/scrapy_crawler/metadata.py:17]: this should be `start_url`

File: hyko_toolkit/functions/web/web_scraping/scrapy_crawler/main.py
Comment (1hachem) [inline hyko_toolkit/functions/web/web_scraping/scrapy_crawler/main.py:1]: maaan this file is a big mess, please try to refactor it and please add docstrings and comments

**Comment by 1hachem** [comment]
I tried to run this function but doesn't seem to work no matter how I try

image

**Comment by ayoubmrx** [comment]
Scrapy might not be compatible with all websites due to variations in site structure and restrictions. It's essential to verify compatibility and adhere to website-specific terms of service for successful crawling.

Screenshot from 2024-05-11 08-53-05

File: hyko_toolkit/functions/web/web_scraping/scrapy_crawler/metadata.py (hunk 1 above)
Comment (ayoubmrx) [reply hyko_toolkit/functions/web/web_scraping/scrapy_crawler/metadata.py:29]: Passing a single domain or a list of domains is mandatory; making it optional would result in errors.

File: hyko_toolkit/functions/web/web_scraping/scrapy_crawler/main.py
Comment (ayoubmrx) [reply hyko_toolkit/functions/web/web_scraping/scrapy_crawler/main.py:1]: I've added comments throughout the code to improve readability and understanding. Regarding the "ReactorNotRestartable" error, the solution I found and implemented in the `run_crawler` function is from this Stack Overflow post: [Scrapy - Reactor not restartable](https://stackoverflow.com/questions/41495052/scrapy-reactor-not-restartable).

**Pull request merged**